From 4b901f6ace9d472e6a1db09335b84c44f774d04b Mon Sep 17 00:00:00 2001
From: Volodymyr Babchuk <volodymyr_babchuk@epam.com>
Date: Fri, 2 Sep 2022 01:06:43 +0300
Subject: [PATCH] plat: rcar: fix core pos calculation for H3 boards

Due to mistake, cluster position wasn't shifted left if chip is not M3W. This
led to erroneous core ID calculation on chips that are not M3W. Actually, this
affected only H3, as only this chip has two clusters.

Fix this by always shifting x1 (cluster ID) to the left, before doing one
additional shift for non-M3W chips.

Fixes: 572afdce53ea ("plat: rcar: Derive core map from PRR")

Signed-off-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>
---
 core/arch/arm/plat-rcar/core_pos_a64.S | 5 +++--
 1 file changed, 3 insertions(+), 2 deletions(-)

diff --git a/core/arch/arm/plat-rcar/core_pos_a64.S b/core/arch/arm/plat-rcar/core_pos_a64.S
index 890f87859..ff25c85e4 100644
--- a/core/arch/arm/plat-rcar/core_pos_a64.S
+++ b/core/arch/arm/plat-rcar/core_pos_a64.S
@@ -64,9 +64,10 @@ FUNC get_core_pos_mpidr , :
 1:	mov	w2, #PRR_PRODUCT_M3W
 	and	w3, w3, #PRR_PRODUCT_MASK
 	cmp	w2, w3
-	bne	2f	/* if (IsM3W) { x1 <<= 1; } */
+	beq	2f	/* if (!IsM3W) { x1 <<= 2; } else { x1 <<= 1} */
 	lsl	x1, x1, #1
-2:	add	x0, x0, x1
+2:	lsl	x1, x1, #1
+	add	x0, x0, x1
 
 	ret
 END_FUNC get_core_pos_mpidr
-- 
2.36.1

